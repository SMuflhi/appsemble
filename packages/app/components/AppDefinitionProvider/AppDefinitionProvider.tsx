import { AppDefinition, BlockManifest } from '@appsemble/types';
import * as React from 'react';

import resolveJsonPointers from '../../utils/resolveJsonPointers';
import settings from '../../utils/settings';

interface AppDefinitionContext {
  blockManifests: BlockManifest[];
  definition: AppDefinition;
  revision: number;
}

interface AppDefinitionProviderProps {
  children: React.ReactNode;
}

function replaceStyle(id: string, style: string): void {
  const oldNode = document.getElementById(id);
  const newNode = document.createElement('style');
  newNode.appendChild(document.createTextNode(style));
  newNode.id = id;

  if (oldNode) {
    document.head.replaceChild(newNode, oldNode);
  } else {
    document.head.appendChild(newNode);
  }
}

const Context = React.createContext<AppDefinitionContext>(null);

export default function AppDefinitionProvider({
  children,
}: AppDefinitionProviderProps): React.ReactElement {
  const [blockManifests, setBlockManifests] = React.useState(settings.blockManifests);
  const [definition, setDefinition] = React.useState(settings.definition);
  const [revision, setRevision] = React.useState(0);

  const value = React.useMemo(() => ({ blockManifests, definition, revision }), [
    blockManifests,
    definition,
    revision,
  ]);

  React.useEffect(() => {
    const onMessage = ({ data }: MessageEvent): void => {
      if (data?.type === 'editor/EDIT_SUCCESS') {
        replaceStyle('appsemble-style-core', data.style);
        replaceStyle('appsemble-style-shared', data.sharedStyle);
        setBlockManifests(data.blockManifests);
        setDefinition(resolveJsonPointers(data.definition) as AppDefinition);
        setRevision(revision + 1);
      }
    };

    window.addEventListener('message', onMessage);

    return () => {
      window.removeEventListener('message', onMessage);
    };
  }, [revision]);

  return <Context.Provider value={value}>{children}</Context.Provider>;
}

export function useAppDefinition(): AppDefinitionContext {
  return React.useContext(Context);
}
