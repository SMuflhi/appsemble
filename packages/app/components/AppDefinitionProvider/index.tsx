import { useEventListener } from '@appsemble/react-components';
import { type AppDefinition, type BlockManifest } from '@appsemble/types';
import { createContext, type ReactNode, useCallback, useContext, useMemo, useState } from 'react';

import {
  apiUrl,
  demoMode,
  blockManifests as initialBlockManifests,
  definition as initialDefinition,
} from '../../utils/settings.js';

interface AppDefinitionContext {
  blockManifests: BlockManifest[];
  definition: AppDefinition;
  demoMode: boolean;
  revision: number;
}

interface AppDefinitionProviderProps {
  readonly children: ReactNode;
}

function replaceStyle(id: string, style: string): void {
  const oldNode = document.getElementById(id);
  const newNode = document.createElement('style');
  newNode.append(document.createTextNode(style));
  newNode.id = id;

  if (oldNode) {
    document.head.replaceChild(newNode, oldNode);
  } else {
    document.head.append(newNode);
  }
}

const Context = createContext<AppDefinitionContext>({
  definition: initialDefinition,
  blockManifests: initialBlockManifests,
  demoMode,
  revision: 0,
});

export function AppDefinitionProvider({ children }: AppDefinitionProviderProps): ReactNode {
  const [blockManifests, setBlockManifests] = useState(initialBlockManifests);
  const [definition, setDefinition] = useState(initialDefinition);
  const [revision, setRevision] = useState(0);

  const value = useMemo(
    () => ({ blockManifests, definition, revision, demoMode }),
    [blockManifests, definition, revision],
  );

  const onMessage = useCallback(
    ({ data, origin }: MessageEvent) => {
      if (origin === apiUrl && data?.type === 'editor/EDIT_SUCCESS') {
        replaceStyle('appsemble-style-core', data.coreStyle);
        replaceStyle('appsemble-style-shared', data.sharedStyle);
        setBlockManifests(data.blockManifests);
        setDefinition(data.definition);
        setRevision(revision + 1);
      } else if (origin === apiUrl && data?.type === 'editor/gui/EDIT_SUCCESS') {
        replaceStyle('appsemble-style-core', data.coreStyle);
        replaceStyle('appsemble-style-shared', data.sharedStyle);
        setDefinition(data.definition);
        setRevision(revision + 1);
      }
    },
    [revision],
  );

  useEventListener(window, 'message', onMessage);

  return <Context.Provider value={value}>{children}</Context.Provider>;
}

export function useAppDefinition(): AppDefinitionContext {
  return useContext(Context);
}
