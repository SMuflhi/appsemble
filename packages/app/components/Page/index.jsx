import { connect } from 'react-redux';

import { getBlockDefs } from '../../actions/blockDefs';
import Page from './Page';

function mapStateToProps(state, ownProps) {
  return {
    app: state.app.app,
    user: state.user.user,
    hasErrors: ownProps.page.blocks.some(block => state.blockDefs.errored.has(block.type)),
  };
}

export default connect(
  mapStateToProps,
  {
    getBlockDefs,
  },
)(Page);
