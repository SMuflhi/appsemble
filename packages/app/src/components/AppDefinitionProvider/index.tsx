import { useEventListener } from '@appsemble/react-components';
import type { AppDefinition, BlockManifest } from '@appsemble/types';
import React, {
  createContext,
  ReactElement,
  ReactNode,
  useCallback,
  useContext,
  useMemo,
  useState,
} from 'react';

import { resolveJsonPointers } from '../../utils/resolveJsonPointers';
import {
  apiUrl,
  blockManifests as initialBlockManifests,
  definition as initialDefinition,
} from '../../utils/settings';

interface AppDefinitionContext {
  blockManifests: BlockManifest[];
  definition: AppDefinition;
  revision: number;
}

interface AppDefinitionProviderProps {
  children: ReactNode;
}

function replaceStyle(id: string, style: string): void {
  const oldNode = document.getElementById(id);
  const newNode = document.createElement('style');
  newNode.append(document.createTextNode(style));
  newNode.id = id;

  if (oldNode) {
    document.head.replaceChild(newNode, oldNode);
  } else {
    document.head.append(newNode);
  }
}

const Context = createContext<AppDefinitionContext>(null);

export function AppDefinitionProvider({ children }: AppDefinitionProviderProps): ReactElement {
  const [blockManifests, setBlockManifests] = useState(initialBlockManifests);
  const [definition, setDefinition] = useState(initialDefinition);
  const [revision, setRevision] = useState(0);

  const value = useMemo(() => ({ blockManifests, definition, revision }), [
    blockManifests,
    definition,
    revision,
  ]);

  const onMessage = useCallback(
    ({ data, origin }: MessageEvent) => {
      if (origin === apiUrl && data?.type === 'editor/EDIT_SUCCESS') {
        replaceStyle('appsemble-style-core', data.style);
        replaceStyle('appsemble-style-shared', data.sharedStyle);
        setBlockManifests(data.blockManifests);
        setDefinition(resolveJsonPointers(data.definition) as AppDefinition);
        setRevision(revision + 1);
      }
    },
    [revision],
  );

  useEventListener(window, 'message', onMessage);

  return <Context.Provider value={value}>{children}</Context.Provider>;
}

export function useAppDefinition(): AppDefinitionContext {
  return useContext(Context);
}
