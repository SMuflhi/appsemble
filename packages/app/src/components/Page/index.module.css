.root {
  display: flex;
  flex-flow: column;
  height: 100%;
}

.hasBottomNavigation > :last-child {
  padding-bottom: 7rem;
}

.twitter {
  background-color: hsl(203, 89%, 53%);
  color: hsl(0, 0%, 100%);
  &:hover {
    color: hsl(0, 0%, 100%);
  }
}

.facebook {
  background-color: hsl(220, 46%, 48%);
  color: hsl(0, 0%, 100%);
  &:hover {
    color: hsl(0, 0%, 100%);
  }
}
