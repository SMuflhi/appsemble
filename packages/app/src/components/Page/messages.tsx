import { defineMessages } from 'react-intl';

export default defineMessages({
  permissionError:
    'Due to misconfigured permissions, youâ€™re not able to view the app. If you think this is a mistake, please leave a message in the <a>Appsemble store page</a>.',
  logout: 'Logout',
});
