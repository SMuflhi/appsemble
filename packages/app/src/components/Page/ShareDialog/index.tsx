import { Button, ModalCard } from '@appsemble/react-components';
import { Dispatch, ReactElement, SetStateAction, useCallback } from 'react';
import { FormattedMessage } from 'react-intl';

import { ShowShareDialogParams } from '../../../types';
import styles from './index.module.css';
import { messages } from './messages';

export interface ShareDialogParams {
  shareDialogParams: ShareDialogState;
  setShareDialogParams: Dispatch<SetStateAction<ShareDialogState>>;
}

export interface ShareDialogState {
  params: ShowShareDialogParams;
  resolve: () => void;
  reject: (error: string) => void;
}

export function ShareDialog({
  setShareDialogParams,
  shareDialogParams,
}: ShareDialogParams): ReactElement {
  const rejectShareDialog = useCallback(() => {
    setShareDialogParams((old) => {
      old?.reject('Closed share dialog');
      return null;
    });
  }, [setShareDialogParams]);
  const resolveShareDialog = useCallback(() => {
    setShareDialogParams((old) => {
      old?.resolve();
      return null;
    });
  }, [setShareDialogParams]);

  return (
    <ModalCard
      isActive={Boolean(shareDialogParams)}
      onClose={rejectShareDialog}
      title={<FormattedMessage {...messages.share} />}
    >
      <div className="buttons is-justify-content-center">
        <Button
          component="a"
          href="mailto:?subject=Test Subject&body=Body!"
          icon="envelope"
          onClick={resolveShareDialog}
        >
          <FormattedMessage {...messages.email} />
        </Button>
        <Button
          className={styles.twitter}
          component="a"
          href={`https://twitter.com/intent/tweet?text=Hello world&url=${window.origin}`}
          icon="twitter"
          onClick={resolveShareDialog}
          rel="noopener noreferrer"
          target="_blank"
        >
          <FormattedMessage {...messages.shareOn} values={{ name: 'Twitter' }} />
        </Button>
        <Button
          className={styles.facebook}
          component="a"
          href={`https://www.facebook.com/sharer/sharer.php?u=${window.origin}&t=ExampleTitle`}
          icon="facebook-f"
          onClick={resolveShareDialog}
          rel="noopener noreferrer"
          target="_blank"
        >
          <FormattedMessage {...messages.shareOn} values={{ name: 'Facebook' }} />
        </Button>
      </div>
    </ModalCard>
  );
}
