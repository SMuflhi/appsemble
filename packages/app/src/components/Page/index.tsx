import { Button, Content, Message, useLocationString } from '@appsemble/react-components';
import type { PageDefinition } from '@appsemble/types';
import { checkAppRole, normalize } from '@appsemble/utils';
import { EventEmitter } from 'events';
import React, { ReactElement, useCallback, useEffect, useRef, useState } from 'react';
import { FormattedMessage } from 'react-intl';
import { Redirect, Route, Switch, useRouteMatch } from 'react-router-dom';

import type { ShowDialogParams } from '../../types';
import settings from '../../utils/settings';
import { useAppDefinition } from '../AppDefinitionProvider';
import BlockList from '../BlockList';
import FlowPage from '../FlowPage';
import PageDialog from '../PageDialog';
import TabsPage from '../TabsPage';
import TitleBar from '../TitleBar';
import { useUser } from '../UserProvider';
import styles from './index.css';
import messages from './messages';

export default function Page(): ReactElement {
  const { definition } = useAppDefinition();
  const redirect = useLocationString();
  const { isLoggedIn, role } = useUser();
  const {
    params: { lang, pageId },
    path,
    url,
  } = useRouteMatch<{ lang: string; pageId: string }>();

  const [dialog, setDialog] = useState<ShowDialogParams>();

  const ee = useRef<EventEmitter>();
  if (!ee.current) {
    ee.current = new EventEmitter();
  }

  const index = definition.pages.findIndex((p) => normalize(p.name) === pageId);
  const page = index === -1 ? null : definition.pages[index];
  const prefix = index === -1 ? null : `pages.${index}`;

  const showDialog = useCallback((d: ShowDialogParams) => {
    setDialog(d);
    return () => {
      setDialog(null);
    };
  }, []);

  useEffect(() => {
    if (!page) {
      return;
    }
    const queryStringParams = new URLSearchParams({ ...definition.theme, ...page.theme });
    const bulmaStyle = document.getElementById('bulma-style-app') as HTMLLinkElement;
    const bulmaUrl = new URL(bulmaStyle.href);
    queryStringParams.sort();
    bulmaUrl.search = String(queryStringParams);
    bulmaStyle.href = String(bulmaUrl);
  }, [definition, page]);

  useEffect(() => {
    if (ee.current) {
      ee.current.removeAllListeners();
      ee.current = null;
      ee.current = new EventEmitter();
    }

    return () => {
      if (ee.current) {
        ee.current.removeAllListeners();
        ee.current = null;
      }
    };
  }, [page]);

  const checkPagePermissions = (p: PageDefinition): boolean => {
    const roles = p.roles || definition.roles || [];
    return roles.length === 0 || roles.some((r) => checkAppRole(definition.security, r, role));
  };

  // If the user is on an existing page and is allowed to view it, render it.
  if (page && checkPagePermissions(page)) {
    return (
      <main className={styles.root} data-path={prefix}>
        <TitleBar>{page.name}</TitleBar>
        {page.type === 'tabs' ? (
          <TabsPage
            ee={ee.current}
            page={page}
            prefix={prefix}
            showDialog={showDialog}
            subPages={page.subPages}
          />
        ) : (
          // The switch is used to enforce an exact path.
          <Switch>
            <Route exact path={path}>
              {page.type === 'flow' ? (
                <FlowPage
                  definition={definition}
                  ee={ee.current}
                  page={page}
                  prefix={prefix}
                  showDialog={showDialog}
                />
              ) : (
                <BlockList
                  blocks={page.blocks}
                  ee={ee.current}
                  page={page}
                  prefix={`${prefix}.blocks`}
                  showDialog={showDialog}
                />
              )}
            </Route>
            {/* Redirect from a matching sub URL to the actual URL */}
            <Redirect to={url} />
          </Switch>
        )}
        <PageDialog dialog={dialog} ee={ee.current} page={page} showDialog={showDialog} />
      </main>
    );
  }

  // If the user isn’t allowed to view the page, because they aren’t logged in, redirect to the
  // login page.
  if (!isLoggedIn) {
    return <Redirect to={`/${lang}/Login?${new URLSearchParams({ redirect })}`} />;
  }

  // If the user is logged in, but isn’t allowed to view the current page, redirect to the default
  // page.
  const defaultPage = definition.pages.find((p) => p.name === definition.defaultPage);
  if (checkPagePermissions(defaultPage)) {
    return <Redirect to={`/${lang}/${normalize(defaultPage.name)}`} />;
  }

  // If the user isn’t allowed to view the default page either, find a page to redirect the user to.
  const redirectPage = definition.pages.find(checkPagePermissions);
  if (redirectPage) {
    return <Redirect to={`/${lang}/${normalize(redirectPage.name)}`} />;
  }

  // If the user isn’t allowed to view any pages, show an error message.
  return (
    <Content padding>
      <Message color="danger">
        <p>
          <FormattedMessage
            {...messages.permissionError}
            values={{
              a: (text: string) => (
                <a
                  href={`${settings.apiUrl}/apps/${settings.id}`}
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  {text}
                </a>
              ),
            }}
          />
        </p>
        <Button className="mt-4" color="danger">
          <FormattedMessage {...messages.logout} />
        </Button>
      </Message>
    </Content>
  );
}
