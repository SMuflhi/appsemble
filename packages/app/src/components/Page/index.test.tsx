import type { BasicPage, FlowPage, TabsPage, UserInfo } from '@appsemble/types';
import { shallow } from 'enzyme';
import React from 'react';

import * as AppDefinitionProvider from '../AppDefinitionProvider';
import Page from '.';

jest.mock('events', () => ({
  EventEmitter: class EventEmitter {},
}));

jest.mock('react-router-dom', () => ({
  useLocation: () => ({
    pathname: '/my-page',
    search: '',
    hash: '',
  }),
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

jest.mock('react-intl', () => {
  const reactIntl = jest.requireActual('react-intl');
  const intl = reactIntl.createIntl({
    locale: 'en',
  });

  return {
    ...reactIntl,
    useIntl: () => intl,
  };
});

jest.mock('../UserProvider', () => ({
  useUser: () => ({
    logout: jest.fn(),
    role: null as string,
    userInfo: null as UserInfo,
  }),
}));

beforeEach(() => {
  jest.spyOn(document, 'getElementById').mockReturnValue(document.createElement('link'));
});

it('should render the blocks for a page', () => {
  const page: BasicPage = {
    name: 'Test Page',
    blocks: [
      { type: 'test', version: '0.0.0' },
      { type: 'test', version: '0.0.0' },
    ],
  };
  jest.spyOn(AppDefinitionProvider, 'useAppDefinition').mockReturnValue({
    revision: 1,
    blockManifests: [],
    definition: {
      defaultPage: 'Test Page',
      pages: [page],
    },
  });
  const wrapper = shallow(<Page page={page} />);

  expect(wrapper).toMatchSnapshot();
});

it('should render tabs pages', () => {
  const page: TabsPage = {
    name: 'Test Page',
    type: 'tabs',
    subPages: [
      {
        name: 'Sub A',
        blocks: [
          { type: 'test', version: '0.0.0' },
          { type: 'test', version: '0.0.0' },
        ],
      },
      {
        name: 'Sub B',
        blocks: [
          { type: 'test', version: '0.0.0' },
          { type: 'test', version: '0.0.0' },
        ],
      },
    ],
  };
  jest.spyOn(AppDefinitionProvider, 'useAppDefinition').mockReturnValue({
    revision: 1,
    blockManifests: [],
    definition: {
      defaultPage: 'Test Page',
      pages: [page],
    },
  });
  const wrapper = shallow(<Page page={page} />);

  expect(wrapper).toMatchSnapshot();
});

it('should render flow page', () => {
  const page: FlowPage = {
    name: 'Test Page',
    type: 'flow',
    subPages: [
      {
        name: 'Sub A',
        blocks: [
          { type: 'test', version: '0.0.0' },
          { type: 'test', version: '0.0.0' },
        ],
      },
      {
        name: 'Sub B',
        blocks: [
          { type: 'test', version: '0.0.0' },
          { type: 'test', version: '0.0.0' },
        ],
      },
    ],
  };
  jest.spyOn(AppDefinitionProvider, 'useAppDefinition').mockReturnValue({
    revision: 1,
    blockManifests: [],
    definition: {
      defaultPage: 'Test Page',
      pages: [page],
    },
  });
  const wrapper = shallow(<Page page={page} />);

  expect(wrapper).toMatchSnapshot();
});
