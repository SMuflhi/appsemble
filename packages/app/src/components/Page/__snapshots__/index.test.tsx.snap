// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render flow page 1`] = `
<Fragment>
  <TitleBar>
    Test Page
  </TitleBar>
  <FlowPage
    definition={
      Object {
        "defaultPage": "Test Page",
        "pages": Array [
          Object {
            "name": "Test Page",
            "subPages": Array [
              Object {
                "blocks": Array [
                  Object {
                    "type": "test",
                    "version": "0.0.0",
                  },
                  Object {
                    "type": "test",
                    "version": "0.0.0",
                  },
                ],
                "name": "Sub A",
              },
              Object {
                "blocks": Array [
                  Object {
                    "type": "test",
                    "version": "0.0.0",
                  },
                  Object {
                    "type": "test",
                    "version": "0.0.0",
                  },
                ],
                "name": "Sub B",
              },
            ],
            "type": "flow",
          },
        ],
      }
    }
    ee={EventEmitter {}}
    page={
      Object {
        "name": "Test Page",
        "subPages": Array [
          Object {
            "blocks": Array [
              Object {
                "type": "test",
                "version": "0.0.0",
              },
              Object {
                "type": "test",
                "version": "0.0.0",
              },
            ],
            "name": "Sub A",
          },
          Object {
            "blocks": Array [
              Object {
                "type": "test",
                "version": "0.0.0",
              },
              Object {
                "type": "test",
                "version": "0.0.0",
              },
            ],
            "name": "Sub B",
          },
        ],
        "type": "flow",
      }
    }
    prefix="pages.0"
    showDialog={[Function]}
  />
  <PageDialog
    ee={EventEmitter {}}
    showDialog={[Function]}
  />
</Fragment>
`;

exports[`should render tabs pages 1`] = `
<Fragment>
  <TitleBar>
    Test Page
  </TitleBar>
  <TabsPage
    ee={EventEmitter {}}
    prefix="pages.0"
    showDialog={[Function]}
    subPages={
      Array [
        Object {
          "blocks": Array [
            Object {
              "type": "test",
              "version": "0.0.0",
            },
            Object {
              "type": "test",
              "version": "0.0.0",
            },
          ],
          "name": "Sub A",
        },
        Object {
          "blocks": Array [
            Object {
              "type": "test",
              "version": "0.0.0",
            },
            Object {
              "type": "test",
              "version": "0.0.0",
            },
          ],
          "name": "Sub B",
        },
      ]
    }
  />
  <PageDialog
    ee={EventEmitter {}}
    showDialog={[Function]}
  />
</Fragment>
`;

exports[`should render the blocks for a page 1`] = `
<Fragment>
  <TitleBar>
    Test Page
  </TitleBar>
  <BlockList
    blocks={
      Array [
        Object {
          "type": "test",
          "version": "0.0.0",
        },
        Object {
          "type": "test",
          "version": "0.0.0",
        },
      ]
    }
    ee={EventEmitter {}}
    prefix="pages.0.blocks"
    showDialog={[Function]}
  />
  <PageDialog
    ee={EventEmitter {}}
    showDialog={[Function]}
  />
</Fragment>
`;
